python/ray/experimental/tf_utils.py

```diff
@@ -195,3 +200,353 @@ def set_weights(self, new_weights):
                 for (name, value) in new_weights.items()
                 if name in self.placeholders
             })
+
+
+DEBUG = 0
+def logger_debug_info(txt):
+    if DEBUG:
+        logger.info(txt)
+
+banner_half = "".join(["#"] * 10)
+
+def tf_differentiable(num_return_vals):
+    def tf_differentiable_helper(method):
+        #TODO(vsatish): Can we automatically determine 
+        #that a method is differentiable at runtime based on inputs?
+
+#        def differentiable_method(self, identifier, forward_pass, 
+#                                  persistent_tape, dys,
+#                                  arg_types, kwarg_types, *args, 
+#                                  **kwargs):
+        def differentiable_method(self, identifier, forward_pass, 
+                                  persistent_tape, dys,
+                                  arg_types, *args):
+            logger_debug_info("{0} Executing wrapper for remote "
+                              "differentiable function '{1}' {0}".format(banner_half, method.__name__))
+            
+
+            if not tf.executing_eagerly():
+                #TODO(vsatish): Can we automatically 
+                #enable this? (Tricky b/c it must be done at start of program.)
+                raise RuntimeError("Tensorflow Eager execution must "
+                                   "be enabled for experimental differentiable ray " 
+                                   "functionality.")
+
+            if forward_pass:
```

Comment:
The forward pass and backward pass should be implemented in two separate helper methods.